python/cudf_polars/cudf_polars/experimental/explain.py

```diff
@@ -0,0 +1,230 @@
+    *,
+    offset: str = "",
+) -> str:
+    count = partition_info[ir].count if partition_info else None
+    header = f"{offset}{type(ir).__name__.upper()}"
+    if count is not None:
+        header += f" [{count}]"
+    return header + "\n"
+
+
+@_format_ir.register
+def _(
+    ir: GroupBy,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    keys = tuple(ne.name for ne in ir.keys)
+    count = partition_info[ir].count if partition_info else None
+    return (
+        f"{offset}GROUPBY {keys}" + (f" [{count}]" if count is not None else "") + "\n"
+    )
+
+
+@_format_ir.register
+def _(
+    ir: Join,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    left_on = tuple(ne.name for ne in ir.left_on)
+    right_on = tuple(ne.name for ne in ir.right_on)
+    count = partition_info[ir].count if partition_info else None
+    return (
+        f"{offset}JOIN {ir.options[0]} {left_on} {right_on}"
+        + (f" [{count}]" if count is not None else "")
+        + "\n"
+    )
+
+
+@_format_ir.register
+def _(
+    ir: Projection,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    count = partition_info[ir].count if partition_info else None
+    return (
+        f"{offset}PROJECTION {tuple(ir.schema)}"
+        + (f" [{count}]" if count is not None else "")
+        + "\n"
+    )
+
+
+@_format_ir.register
+def _(
+    ir: Select,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    count = partition_info[ir].count if partition_info else None
+    return (
+        f"{offset}SELECT {tuple(ir.schema)}"
+        + (f" [{count}]" if count is not None else "")
+        + "\n"
+    )
+
+
+@_format_ir.register
+def _(
+    ir: Sort,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    by = tuple(ne.name for ne in ir.by)
+    count = partition_info[ir].count if partition_info else None
+    return f"{offset}SORT {by}" + (f" [{count}]" if count is not None else "") + "\n"
+
+
+def _format_scan_summary(
+    typ: str,
+    schema: tuple,
+    paths: list[str],
+    *,
+    count: int | None = None,
+    prefix: str = "",
+) -> str:
+    first_path = paths[0]
+    suffix = " ..." if len(paths) > 1 else ""
+    header = f"{prefix}{typ.upper()} {schema} {first_path}{suffix}"
+    if count is not None:
+        header += f" [{count} partition{'s' if count > 1 else ''}]"
+    return header
+
+
+@_format_ir.register
+def _(
+    ir: Union,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    count = partition_info[ir].count if partition_info else None
+
+    if ir.children and isinstance(ir.children[0], (Scan, SplitScan)):
```

Comment:
I'm not sure that we need to do any special handling for `Union` now that you are handling repeated children in `_format_ir_tree`. The original code in `pdsh.py` was only doing something like this to avoid that annoying repetition.